### PR TITLE
[WIP] Enable causal block mask for sdpa

### PR DESCRIPTION
**Pull request by mreso**
This PR enables causal block mask for sdpa using nested tensors.

To test run:
```
CONFIG_FILE="./torchtitan/models/llama3/train_configs/debug_model.toml" ./run_train.sh --training.local_batch_size=2 --model.flavor=debugmodel_sdpa_block_causal
```

TODO
[X] Test with debug model
[] Test with llama3 8B
[] Test with llama3 70B

### CODE REVIEW

**Comment by fegin** [comment]
I thought we want to pursue SDPA with variable length support instead of using NestedTensor as the uncertainty of DTensor + NestedTensor composability issue. We didn't investigate how to do CP with SDPA + NestedTensor because CP + SDPA relies on the DTensor dispatcher.

cc., @drisspg

**Comment by mreso** [comment]
> I thought we want to pursue SDPA with variable length support instead of using NestedTensor as the uncertainty of DTensor + NestedTensor composability issue. We didn't investigate how to do CP with SDPA + NestedTensor because CP + SDPA relies on the DTensor dispatcher.
> 
> cc., @drisspg

Wasn't in this case the variable length support in SDPA and nested tensor just a different kind of interface for the same thing?

**Comment by fegin** [comment]
It's not the same. If SDPA directly supports variable length, that will be the native tensor cases, similar to directly export FlashAttention interface to SDPA without the needed of NestedTensor. NestedTensor + DTensor (and CP) is an uncertainty to us.

**Comment by mreso** [comment]
Closing this as NJT is not composable with DTensor.
For the record, local repro of the last issue we encountered:
```
import os
import torch
import torch.nn.functional as F
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.nested import nested_tensor_from_jagged
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor.experimental import context_parallel
class ScaledDotProductAttention(torch.nn.Module):
    offsets = None
    def __init__(self):
        super().__init__()
        ScaledDotProductAttention.offsets=torch.tensor(
            [    0,   161,   542,   587,   825,  1020,  1143,  1193,  1311,  1437,
             1886,  1962,  2256,  2511,  2608,  2658,  5097,  6016,  6440,  6798,
             8428,  8658,  8856,  9187,  9406,  9683, 10084, 10478, 11125, 11505,
             11774, 12698, 13921, 13974, 14011, 14174, 14353, 14722, 14808, 15098,
            15600, 15696, 16026, 16183, 16767, 16822, 16884, 17146, 17294, 17732,
            17931, 18415, 18788, 18889, 19199, 19331, 19523, 19653, 20778, 21239,
            21857, 22408, 22674, 22805, 28190, 28286, 28606, 28970, 29512, 29683,
            30334, 30407, 30494, 32179, 32481, 32649, 32768], device='cuda:0')
        
    def forward(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor):
        original_shape = q.shape
        if q.size(0) == 1:
            # Create nested tensor: [1, h, s, d] -> [num_samples, h, j1, d]
            
            q_nested = nested_tensor_from_jagged(
                q.view(q.shape[1:]),
                ScaledDotProductAttention.offsets,
                jagged_dim=2,
                min_seqlen=1,
                max_seqlen=8192,
            )
            k_nested = nested_tensor_from_jagged(
                k.view(k.shape[1:]),
                ScaledDotProductAttention.offsets,
                jagged_dim=2,
                min_seqlen=1,
                max_seqlen=8192,
            )
            v_nested = nested_tensor_from_jagged(
                v.view(v.shape[1:]),
                ScaledDotProductAttention.offsets,
                jagged_dim=2,
                min_seqlen=1,
                max_seqlen=8192,
            )
            act_nested = F.scaled_dot_product_attention(
                q_nested, k_nested, v_nested, is_causal=True
            )
            return act_nested.values().view(original_shape)
        else:
            assert 0
def call_sdpa(rank, world_size):
    torch.cuda.set_device(f'cuda:{rank}')
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'
    print(f"{rank=}")
    dist.init_process_group(backend='nccl', rank=rank, world_size=world_size, device_id=rank)
    cp_mesh = init_device_mesh("cuda", (2,), mesh_dim_names=("cp",))
    torch._dynamo.config.capture_scalar_outputs = True
    q = torch.randint(0, 100, size=(1, 16, 32768, 16), dtype=torch.bfloat16, device='cuda:0')
    k = torch.randint(0, 100, size=(1, 16, 32768, 16), dtype=torch.bfloat16, device='cuda:0')
    v = torch.randint(0, 100, size=(1, 16, 32768, 16), dtype=torch.bfloat16, device='cuda:0')
        
    with context_parallel(cp_mesh,  buffers=(q,k,v), buffer_seq_dims=(2, 2, 2)):
        
        module = ScaledDotProductAttention()
        
        compiled_module = torch.compile(module, fullgraph=True)
        
        act = compiled_module(q,k,v)
    dist.barrier()
    dist.destroy_process_group()
def test_sdpa():
    mp.spawn(call_sdpa, args=(2,), nprocs=2, join=True)
if __name__ == "__main__":
    test_sdpa()
```
